graph/structure/impl/RuleStructureImpl.java

```diff
@@ -91,6 +102,26 @@ public boolean isDeleted() {
         return isDeleted.get();
     }
 
+    @Override
+    public void createConcludingIsaIndex(Label type) {
+        graph.ruleIndex().concludingIsa(this, type);
+    }
+
+    @Override
+    public void clearConcludingIsaIndex(Label type) {
+        graph.ruleIndex().deleteConcludingIsa(this, type);
+    }
+
+    @Override
+    public void createConcludingHasAttributeIndex(Label type) {
+        graph.ruleIndex().concludingHasAttribute(this, type);
+    }
+
+    @Override
+    public void clearConcludesHasAttributeIndex(Label type) {
+        graph.ruleIndex().deleteConcludingHasAttribute(this, type);
```

Comment:
injecting data into the rule index from the higher level reasoning computation - we pass through the `RuleStructure` to the `SchemaGraph.RuleIndex`